File: Semestre 4/IASA_Aux/IASA/2020_2021/tp2/src/tp2/memproc/mem/No.java
package tp2.memproc.mem;

import tp2.modprob.Estado;
import tp2.modprob.Operador;
import tp2.pee.PassoSolucao;

public class No implements PassoSolucao{
	private int profundidade = 0;
	private double custo = 0;
	private Estado estado;
	private Operador operador;
	private No antecessor;
	public No(Estado estado) {
		this.estado = estado;
	}
	
	public No(Estado estado, Operador operador, No antecessor) {
		this.estado = estado;
		this.operador = operador;
		this.antecessor = antecessor;
		
		if(this.antecessor != null) {
			this.profundidade = this.antecessor.profundidade + 1;
			this.custo = operador.custo(antecessor.getEstado(), estado) + antecessor.getCusto();
		}

	}
	
	public Estado getEstado() {
		return this.estado;
		
	}
	public Operador getOperador() {
		return this.operador;
	}
	public No getAntecessor() {
		return this.antecessor;
		
	}
	public int getProfundidade(){
		return this.profundidade;
		
	}
	public double getCusto() {
		return this.custo;
		
	}
	
	@Override
	public String toString() {
		return this.getCusto() + " - " + this.getEstado();
	}

}

File: Semestre 4/IASA/Projecto/iasa_agente/src/lib/pee/mec_proc/no.py
class No:

    def __init__(self, estado, operador=None, antecessor=None):
        self.__estado = estado
        self.__operador = operador
        self.__antecessor = antecessor
        self.__profundidade = 0
        self.__custo = 0

    @property
    def profundidade(self):
        return self.__profundidade

    @property
    def custo(self):
        return self.__custo
    
    @property
    def estado(self):
        return self.__estado
    
    @property
    def operador(self):
        return self.__operador
    
    @property
    def custo(self):
        return self.__antecessor

    def __lt__(self, no):
        return self.__custo < no.custo